src/index.ts

```diff
@@ -72,7 +74,14 @@ export class CSG {
       }
 
       for (let j = 3; j <= pvlen; j++) {
-        const fc = new Face3(v0, v0 + j - 2, v0 + j - 1);
+        const fc = new Face3(
+          v0,
+          v0 + j - 2,
+          v0 + j - 1,
+          undefined,
+          undefined,
+          pvs[0].materialIdx
```

Comment:
The only potential issue I see with this PR is this line, as it takes the materialIndex of the first vertex of the `Face3` to apply to the whole `Face3`. I don't know how the BSP algorithm really works, but I guess it might be possible in some cases that it would create a `Face3` with heterogeneous vertices (i.e. not all the three vertices of the `Face3` after CSG have the same material index as before the CSG).
You might decide to add an additional check which verifies this (i.e. if `pvs[0].materialIdx === pvs[1].materialIdx === pvs[2].materialIdx`) and in the case where the test fails, set the materialIdx to 0 (instead of `pvs[0].materialIdx`), which is what your code was implicitly doing until now.
Your call...